PolyEngine/Engine/Src/ECS/ComponentIterator.hpp

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include "Defines.hpp"
+
+namespace Poly
+{
+	template<typename T>
+	class Dynarray;
+	class ComponentBase;
+	template<typename PrimaryComponent, typename... SecondaryComponents>
+	struct IteratorProxy;
+
+	class IEntityIteratorHelper : public BaseObject<>
+	{
+		public:
+			virtual bool operator==(const IEntityIteratorHelper&) const = 0;
+			virtual bool operator!=(const IEntityIteratorHelper&) const = 0;
+			virtual Entity* get() const = 0;
+			virtual void increment() = 0;
+			virtual bool isValid() const = 0;
+	};
+
+	template<typename PrimaryComponent, typename... SecondaryComponents>
+	class ComponentIterator : public BaseObject<>,
+							public std::iterator<std::forward_iterator_tag, std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type...>>	
+	{
+		public:
+			explicit ComponentIterator(std::unique_ptr<IEntityIteratorHelper> iter) : Iter(std::move(iter))
+			{
+				UpdateCache();
+			}
+			bool operator==(const  ComponentIterator& rhs) const { return Iter.get()->get() == rhs.Get()->get(); } //they never  trigger underlying overload, pointer comparison ongoing
+			bool operator!=(const  ComponentIterator& rhs) const { return !(Iter.get()->get() == rhs.Get()->get()); }
+
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type... > operator*() const //canot be const if  change cache
+			{
+				return GetCache();
+			}
+			std::tuple<typename std::add_pointer<PrimaryComponent>::type, typename std::add_pointer<SecondaryComponents>::type... > operator->() const
+			{
+				return GetCache();
+			}
+
+			ComponentIterator& operator++() { Increment(); return *this; }
+			ComponentIterator operator++(int) { ComponentIterator ret(Iter); Increment(); return ret; } //test for double incrementing etc
+
+			IEntityIteratorHelper* Get() const
```

Comment:
Why do we need Helper to be visible outside of the iterator? This method should be private or protected at most. Name of this method requires more consideration.

Reply:
Changed to GetIteratorHelper and decided to move this to protected (I think that inheriting classes will have no need of reimplementing this method - unless they want to keep pointers to several IEntityIteratorHelpers)